src/lib.rs

```diff
@@ -145,3 +158,52 @@ impl DeviceManager {
         Ok(())
     }
 }
+
+async fn get_device_id(opt_device_id: Option<String>) -> Result<String, DeviceManagerError> {
+    if let Some(device_id) = opt_device_id {
+        Ok(device_id)
+    } else {
+        let connection = zbus::Connection::system().await?;
+        let proxy = DeviceProxy::new(&connection).await?;
+        let namespace = Uuid::parse_str("f79ad91f-c638-4889-ae74-9d001a3b4cf8").unwrap();
+        let hardware_id: String = proxy.device_id().await?;
+        if hardware_id.is_empty() {
+            return Err(DeviceManagerError::FatalError(
+                "No hardware id provided".to_string(),
+            ));
+        }
+        Ok(registration::generate_uuid(namespace, &hardware_id))
+    }
+}
+
+async fn get_credentials_secret(
+    device_id: &str,
+    opts: &DeviceManagerOptions,
+) -> Result<String, DeviceManagerError> {
+    if opts.credentials_secret.is_some() {
+        Ok(opts.credentials_secret.clone().unwrap())
```

Comment:
same as before, this is an `if let` done manually, it's a good practice to avoid calling `.unwrap()` and variants unless absolutely necessary